internal/argocd/revision.go

```diff
@@ -10,68 +11,163 @@ import (
-			// the repo URL is an OCI registry URL. Kargo Warehouses and Freight,
-			// however, do use oci:// at the beginning of such URLs.
-			//
-			// Additionally, where OCI is concerned, an ApplicationSource's repoURL is
-			// really a registry URL, and the chart name is a repository within that
-			// registry. Warehouses and Freight, however, handle things more correctly
-			// where a repoURL points directly to a repository and chart name is
-			// irrelevant / blank. We need to account for this when we search our
-			// Freight for the chart.
-			repoURL = fmt.Sprintf(
-				"oci://%s/%s",
-				strings.TrimSuffix(repoURL, "/"),
-				chartName,
-			)
-			chartName = ""
+
+		if desiredRevision != "" {
+			appLogger.WithValues("source", s,
+				"revision", desiredRevision).Debug("Found desired revision for application source.")
+			return []string{desiredRevision}, nil
 		}
+
+		appLogger.WithValues("source", s).Debug("Could not determine desired revision for application from this source.")
+		return nil, nil
+	}
+
+	// An application may have more than one source that points to the same Git repository,
+	// eg. a Helm and a vanilla manifest source.
+	// In that situation it doesn't matter which source's target revision is returned as ArgoCD does not support
+	// different target revisions for sources targeting the same repository.
+	var revisions = make([]string, len(app.Spec.Sources))
+	isSynced := app.Status.Sync.Revisions != nil
+
+	// ArgoCD uses a shadow-array to store the corresponding revisions preserving the order of app.spec.sources.
+	// We match that logic here and return the desired revision for each source,
+	// or empty string if one is not found.
+	for i, src := range app.Spec.Sources {
+		s := src
+
+		syncedRevisionOfSource := ""
+		if isSynced && i < len(app.Status.Sync.Revisions) {
+			syncedRevisionOfSource = app.Status.Sync.Revisions[i]
+		}
+
+		// If there is a source update that targets app.Spec.Source, it might
+		// have its own ideas about the desired revision.
+		targetUpdate := getTargetUpdate(ctx, update, &s)
+		desiredOrigin := freight.GetDesiredOrigin(ctx, stage, targetUpdate)
+		if desiredOrigin == nil {
+			appLogger.WithValues("source", s,
+				"revision", syncedRevisionOfSource).Debug("Could not determine desired origin" +
+				" for application source, using existing revision.")
+			revisions[i] = syncedRevisionOfSource
+			continue
+		}
+
+		logger.Debug("Resolved origin for application source", "origin", desiredOrigin)
+		desiredRevision, err := getRevisionFromSource(ctx, cl, stage, &s, desiredOrigin, frght)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if desiredRevision != "" {
+			appLogger.Trace("Found desired revision for application source.", "revision", desiredRevision, "source", &s)
+		}
+		revisions[i] = desiredRevision
+	}
+
+	appLogger.Debug("Found desired revisions for multi-source application.", "revisions", revisions)
+	return revisions, nil
+}
+
+func getRevisionFromSource(
```

Comment:
This function name really confused me for a bit as it doesn't seem to accurately describe what it does.

I suggest `getDesiredRevisionForSource()`